runtime/src/accounts_db.rs

```diff
@@ -782,40 +779,29 @@ impl AccountsDB {
         }
     }
 
-    fn handle_reclaims_maybe_cleanup(&self, reclaims: SlotSlice<AccountInfo>) {
-        let mut dead_accounts = Measure::start("reclaims::remove_dead_accounts");
-        let dead_slots = self.remove_dead_accounts(reclaims);
-        dead_accounts.stop();
-        let dead_slots_len = {
-            let mut dead_slots_w = self.dead_slots.write().unwrap();
-            dead_slots_w.extend(dead_slots);
-            dead_slots_w.len()
-        };
-        if dead_slots_len > 5000 {
-            self.process_dead_slots(None);
+    fn handle_reclaims(
```

Comment:
The meaning of the arguments here is not super straightforward. What does expected_slot mean when reclaiming and why can it be `None` and why is cleanup possible/not-possible? Maybe some comments?

Reply:
They were some correctness assertions I added to double check my understanding of the code. Added some comments!

Reply:
@ryoqun, the one I wasn't 100% sure about was this one here in `do_shrink_slot`: https://github.com/solana-labs/solana/pull/12194/files#diff-2099c5256db4eb5975c8834af38f6456L968. I think cleanup can't happen, i.e. it should be safe to mark `is_cleanup_possible == false`, but I erred on the side of caution b/c I'm not 100 sure so I marked `is_cleanup_possible == true`.

My reasoning for why cleanup shouldn't be possible there:

From my understanding right now, the `AccountsBackgroundService` is the only thread that will trigger `shrink_slot() ->  clean_accounts`.

So for cleanup to happen in `shrink_slots`, the `accounts` here: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L963 has to be empty, which means this `slot` has no alive accounts.

However, this is not possible because this check: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L876 guarantees there's at least one alive account, and no other thread is touching this slot between that check and the call to `handle_reclaims`

Reply:
@carllin 

> My reasoning for why cleanup shouldn't be possible there:
> ...

Yeah, your reasoning sounds valid for me. we can call `handle_reclaims(..., None, false)` from `do_shrink_slot`